File: python/compress_zip/arith_coder.py
"""
Arithmetic encoder/decoder for neural compression.

Matches the CUDA/C++ arithmetic coder with 32-bit state.
Uses E1/E2/E3 renormalization scheme.
"""

import numpy as np
from typing import List, Tuple
from io import BytesIO


# Constants matching CUDA implementation
NUM_STATE_BITS = 32
HALF_RANGE = 1 << 31  # 2^31
QUARTER_RANGE = 1 << 30  # 2^30
THREE_QUARTER_RANGE = 3 * QUARTER_RANGE

# Masks
MAX_RANGE = (1 << NUM_STATE_BITS) - 1  # 0xFFFFFFFF


class ArithEncoder:
    """
    Arithmetic encoder with 32-bit state.

    Uses E1/E2/E3 renormalization:
    - E1: low >= HALF, both in upper half -> output 1, then pending 0s
    - E2: high < HALF, both in lower half -> output 0, then pending 1s
    - E3: low in [1/4, 1/2), high in [1/2, 3/4) -> increment pending, rescale

    Bits are packed MSB-first into bytes.
    """

    def __init__(self):
        self.low: int = 0
        self.high: int = MAX_RANGE
        self.pending_bits: int = 0
        self.output_bits: List[int] = []

    def _output_bit(self, bit: int):
        """Output a single bit."""
        self.output_bits.append(bit)

    def _output_bit_with_pending(self, bit: int):
        """Output a bit followed by pending opposite bits."""
        self._output_bit(bit)
        for _ in range(self.pending_bits):
            self._output_bit(1 - bit)
        self.pending_bits = 0

    def encode_symbol(self, cumfreqs: np.ndarray, symbol: int, total: int):
        """
        Encode a single symbol.

        Matches CUDA arith_cuda.cu inclusive cumsum format.

        Args:
            cumfreqs: Cumulative frequency table [vocab_size] (inclusive cumsum)
            symbol: Symbol index to encode (0 to vocab_size - 1)
            total: Total frequency (cumfreqs[-1])
        """
        range_size = self.high - self.low + 1

        # Update range based on symbol (CUDA-matching inclusive format)
        # sym_low = cumfreqs[symbol - 1] if symbol > 0 else 0
        # sym_high = cumfreqs[symbol]
        sym_high = int(cumfreqs[symbol])
        sym_low = int(cumfreqs[symbol - 1]) if symbol > 0 else 0

        # new_high = low + (range * sym_high / total) - 1
        # new_low = low + (range * sym_low / total)
        self.high = self.low + (range_size * sym_high // total) - 1
        self.low = self.low + (range_size * sym_low // total)

        # Renormalization loop
        while True:
            if self.high < HALF_RANGE:
                # E2: Both in lower half
                self._output_bit_with_pending(0)
                self.low = self.low << 1
                self.high = (self.high << 1) | 1
            elif self.low >= HALF_RANGE:
                # E1: Both in upper half
                self._output_bit_with_pending(1)
                self.low = (self.low - HALF_RANGE) << 1
                self.high = ((self.high - HALF_RANGE) << 1) | 1
            elif self.low >= QUARTER_RANGE and self.high < THREE_QUARTER_RANGE:
                # E3: Straddling the middle
                self.pending_bits += 1
                self.low = (self.low - QUARTER_RANGE) << 1
                self.high = ((self.high - QUARTER_RANGE) << 1) | 1
            else:
                break

            # Keep values in range
            self.low &= MAX_RANGE
            self.high &= MAX_RANGE

    def finish(self) -> bytes:
        """
        Finish encoding and return compressed bytes.

        Matches CUDA arith_encode_finish_kernel: emits a single '1' bit as terminator.
        """
        # Emit a fixed '1' bit as terminator (matching CUDA)
        self._output_bit(1)

        # Flush partial byte if any bits written
        if len(self.output_bits) % 8 != 0:
            # Pad remaining bits in current byte with zeros
            while len(self.output_bits) % 8 != 0:
                self.output_bits.append(0)

        # Convert bits to bytes (MSB first)
        result = bytearray()
        for i in range(0, len(self.output_bits), 8):
            byte = 0
            for j in range(8):
                byte = (byte << 1) | self.output_bits[i + j]
            result.append(byte)

        return bytes(result)


class ArithDecoder:
    """
    Arithmetic decoder with 32-bit state.

    Inverse of ArithEncoder.
    """

    def __init__(self, data: bytes):
        self.data = data
        self.bit_pos = 0
        self.low: int = 0
        self.high: int = MAX_RANGE
        self.value: int = 0

        # Initialize value from first 32 bits
        for _ in range(NUM_STATE_BITS):
            self.value = (self.value << 1) | self._read_bit()

    def _read_bit(self) -> int:
        """Read the next bit from the input stream."""
        if self.bit_pos >= len(self.data) * 8:
            return 0  # Pad with zeros

        byte_idx = self.bit_pos // 8
        bit_idx = 7 - (self.bit_pos % 8)  # MSB first
        bit = (self.data[byte_idx] >> bit_idx) & 1
        self.bit_pos += 1
        return bit

    def decode_symbol(self, cumfreqs: np.ndarray, total: int) -> int:
        """
        Decode a single symbol.

        Matches CUDA arith_cuda.cu inclusive cumsum format.

        Args:
            cumfreqs: Cumulative frequency table [vocab_size] (inclusive cumsum)
            total: Total frequency (cumfreqs[-1])

        Returns:
            Decoded symbol index
        """
        range_size = self.high - self.low + 1

        # Find which symbol the current value falls into
        # scaled_value = ((value - low + 1) * total - 1) / range
        scaled = ((self.value - self.low + 1) * total - 1) // range_size

        # Binary search for symbol (CUDA-matching inclusive format)
        # For inclusive cumsum: symbol k spans [cumfreqs[k-1], cumfreqs[k])
        # where cumfreqs[-1] = 0 by convention
        vocab_size = len(cumfreqs)
        symbol = 0
        for i in range(vocab_size):
            if cumfreqs[i] > scaled:
                symbol = i
                break
        else:
            symbol = vocab_size - 1

        # Update range (CUDA-matching inclusive format)
        sym_high = int(cumfreqs[symbol])
        sym_low = int(cumfreqs[symbol - 1]) if symbol > 0 else 0

        self.high = self.low + (range_size * sym_high // total) - 1
        self.low = self.low + (range_size * sym_low // total)

        # Renormalization loop (same as encoder)
        while True:
            if self.high < HALF_RANGE:
                # E2: Both in lower half
                self.low = self.low << 1
                self.high = (self.high << 1) | 1
                self.value = (self.value << 1) | self._read_bit()
            elif self.low >= HALF_RANGE:
                # E1: Both in upper half
                self.low = (self.low - HALF_RANGE) << 1
                self.high = ((self.high - HALF_RANGE) << 1) | 1
                self.value = ((self.value - HALF_RANGE) << 1) | self._read_bit()
            elif self.low >= QUARTER_RANGE and self.high < THREE_QUARTER_RANGE:
                # E3: Straddling the middle
                self.low = (self.low - QUARTER_RANGE) << 1
                self.high = ((self.high - QUARTER_RANGE) << 1) | 1
                self.value = ((self.value - QUARTER_RANGE) << 1) | self._read_bit()
            else:
                break

            # Keep values in range
            self.low &= MAX_RANGE
            self.high &= MAX_RANGE
            self.value &= MAX_RANGE

        return symbol


def encode_symbols(symbols: List[int], cumfreqs: np.ndarray, total: int) -> bytes:
    """Convenience function to encode a list of symbols."""
    encoder = ArithEncoder()
    for sym in symbols:
        encoder.encode_symbol(cumfreqs, sym, total)
    return encoder.finish()


def decode_symbols(data: bytes, cumfreqs: np.ndarray, total: int, num_symbols: int) -> List[int]:
    """Convenience function to decode a list of symbols."""
    decoder = ArithDecoder(data)
    symbols = []
    for _ in range(num_symbols):
        symbols.append(decoder.decode_symbol(cumfreqs, total))
    return symbols


# Test cases
if __name__ == "__main__":
    # Test basic encode/decode roundtrip
    # Simple frequency table: 4 symbols with frequencies [1, 2, 3, 4]
    # Using CUDA-matching inclusive cumsum format (no leading 0)
    freqs = np.array([1, 2, 3, 4], dtype=np.uint32)
    cumfreqs = np.cumsum(freqs).astype(np.uint32)  # Inclusive: [1, 3, 6, 10]
    total = int(cumfreqs[-1])

    print(f"Frequencies: {freqs}")
    print(f"Cumfreqs (inclusive): {cumfreqs}")
    print(f"Total: {total}")

    # Encode some symbols
    original_symbols = [0, 1, 2, 3, 3, 2, 1, 0, 2, 2]
    encoded = encode_symbols(original_symbols, cumfreqs, total)
    print(f"Original: {original_symbols}")
    print(f"Encoded: {len(encoded)} bytes: {encoded.hex()}")

    # Decode
    decoded = decode_symbols(encoded, cumfreqs, total, len(original_symbols))
    print(f"Decoded: {decoded}")

    assert original_symbols == decoded, "Roundtrip failed!"
    print("Roundtrip test passed!")

    # Test with larger vocabulary
    vocab_size = 100
    np.random.seed(42)
    freqs_large = np.random.randint(1, 100, vocab_size, dtype=np.uint32)
    cumfreqs_large = np.cumsum(freqs_large).astype(np.uint32)  # Inclusive
    total_large = int(cumfreqs_large[-1])

    # Generate random symbols
    original_large = [int(np.random.randint(0, vocab_size)) for _ in range(1000)]

    encoded_large = encode_symbols(original_large, cumfreqs_large, total_large)
    decoded_large = decode_symbols(encoded_large, cumfreqs_large, total_large, len(original_large))

    assert original_large == decoded_large, "Large vocab roundtrip failed!"
    print(f"Large vocab test passed! {len(original_large)} symbols -> {len(encoded_large)} bytes")
    print(f"Compression ratio: {len(original_large) / len(encoded_large):.2f}x")

    print("\nAll arithmetic coder tests passed!")

File: rust/src/arith_coder.rs
//! Arithmetic encoder and decoder with 32-bit state.
//!
//! Matches the CUDA implementation in arith_cuda.cu.

use crate::softmax_cdf::{HALF_RANGE, NUM_STATE_BITS, QUARTER_RANGE, STATE_MASK};

/// Arithmetic encoder state.
pub struct ArithEncoder {
    pub low: u64,
    pub high: u64,
    pub underflow: u32,
    /// Output bitstream (packed bytes, MSB first)
    pub output: Vec<u8>,
    /// Current byte being built
    pub current_byte: u8,
    /// Bit position in current byte (7 = MSB, 0 = LSB)
    pub bit_pos: i32,
}

impl Default for ArithEncoder {
    fn default() -> Self {
        Self::new()
    }
}

impl ArithEncoder {
    /// Create a new encoder with initial state.
    pub fn new() -> Self {
        Self {
            low: 0,
            high: STATE_MASK,
            underflow: 0,
            output: Vec::new(),
            current_byte: 0,
            bit_pos: 7,
        }
    }

    /// Reset encoder state for a new chunk.
    pub fn reset(&mut self) {
        self.low = 0;
        self.high = STATE_MASK;
        self.underflow = 0;
        self.output.clear();
        self.current_byte = 0;
        self.bit_pos = 7;
    }

    /// Emit a single bit to the output stream.
    #[inline]
    fn emit_bit(&mut self, bit: u8) {
        if bit != 0 {
            self.current_byte |= 1 << self.bit_pos;
        }
        self.bit_pos -= 1;

        if self.bit_pos < 0 {
            self.output.push(self.current_byte);
            self.current_byte = 0;
            self.bit_pos = 7;
        }
    }

    /// Emit a bit and all pending underflow bits.
    #[inline]
    fn emit_bit_plus_underflow(&mut self, bit: u8) {
        self.emit_bit(bit);
        let opposite = bit ^ 1;
        for _ in 0..self.underflow {
            self.emit_bit(opposite);
        }
        self.underflow = 0;
    }

    /// Encode a symbol given its cumulative frequency range.
    ///
    /// Arguments:
    /// - cumfreqs: cumulative frequencies array (inclusive prefix sum)
    /// - symbol: the symbol to encode (0-indexed)
    pub fn encode_symbol(&mut self, cumfreqs: &[u32], symbol: usize) {
        let total = cumfreqs[cumfreqs.len() - 1] as u64;

        // Get symbol's frequency range
        let sym_high = cumfreqs[symbol] as u64;
        let sym_low = if symbol > 0 {
            cumfreqs[symbol - 1] as u64
        } else {
            0
        };

        // Update interval
        let range = self.high - self.low + 1;
        self.high = self.low + (sym_high * range / total) - 1;
        self.low = self.low + (sym_low * range / total);

        // Renormalize
        self.renormalize();
    }

    /// Renormalization loop (E1, E2, E3 conditions).
    fn renormalize(&mut self) {
        loop {
            // E1/E2: MSBs match - shift out
            if (self.low ^ self.high) & HALF_RANGE == 0 {
                let out_bit = ((self.low >> (NUM_STATE_BITS - 1)) & 1) as u8;
                self.emit_bit_plus_underflow(out_bit);
                self.low = (self.low << 1) & STATE_MASK;
                self.high = ((self.high << 1) & STATE_MASK) | 1;
            }
            // E3: Underflow condition
            else if (self.low & !self.high & QUARTER_RANGE) != 0 {
                self.underflow += 1;
                self.low = ((self.low << 1) ^ HALF_RANGE) & STATE_MASK;
                self.high = (((self.high ^ HALF_RANGE) << 1) | HALF_RANGE | 1) & STATE_MASK;
            } else {
                break;
            }
        }
    }

    /// Finish encoding and flush remaining bits.
    pub fn finish(&mut self) {
        // Emit terminating bit
        self.emit_bit(1);

        // Flush partial byte
        if self.bit_pos < 7 {
            self.output.push(self.current_byte);
        }
    }

    /// Get the encoded bitstream.
    pub fn get_output(&self) -> &[u8] {
        &self.output
    }
}

/// Arithmetic decoder state.
pub struct ArithDecoder {
    pub low: u64,
    pub high: u64,
    pub code: u64,
    /// Input bitstream
    input: Vec<u8>,
    /// Current byte index
    byte_pos: usize,
    /// Bit position in current byte (7 = MSB, 0 = LSB)
    bit_pos: i32,
}

impl ArithDecoder {
    /// Create a new decoder with the given input.
    pub fn new(input: Vec<u8>) -> Self {
        let mut decoder = Self {
            low: 0,
            high: STATE_MASK,
            code: 0,
            input,
            byte_pos: 0,
            bit_pos: 7,
        };
        decoder.init_code();
        decoder
    }

    /// Initialize code register from first 32 bits.
    fn init_code(&mut self) {
        self.code = 0;
        for _ in 0..NUM_STATE_BITS {
            self.code = (self.code << 1) | self.read_bit() as u64;
        }
    }

    /// Read a single bit from the input stream.
    #[inline]
    fn read_bit(&mut self) -> u8 {
        if self.byte_pos >= self.input.len() {
            return 0; // Pad with zeros
        }

        let bit = (self.input[self.byte_pos] >> self.bit_pos) & 1;
        self.bit_pos -= 1;

        if self.bit_pos < 0 {
            self.byte_pos += 1;
            self.bit_pos = 7;
        }

        bit
    }

    /// Decode the next symbol given cumulative frequencies.
    ///
    /// Returns the decoded symbol (0-indexed).
    pub fn decode_symbol(&mut self, cumfreqs: &[u32]) -> usize {
        let total = cumfreqs[cumfreqs.len() - 1] as u64;
        let range = self.high - self.low + 1;
        let offset = self.code - self.low;

        // Compute value in [0, total)
        let value = ((offset + 1) * total - 1) / range;

        // Binary search for symbol
        let symbol = self.bisect_cumfreqs(cumfreqs, value as u32);

        // Get symbol's frequency range
        let sym_high = cumfreqs[symbol] as u64;
        let sym_low = if symbol > 0 {
            cumfreqs[symbol - 1] as u64
        } else {
            0
        };

        // Update interval (same as encoder)
        self.high = self.low + (sym_high * range / total) - 1;
        self.low = self.low + (sym_low * range / total);

        // Renormalize
        self.renormalize();

        symbol
    }

    /// Binary search to find symbol from cumfreqs.
    fn bisect_cumfreqs(&self, cumfreqs: &[u32], value: u32) -> usize {
        // Find smallest i such that value < cumfreqs[i]
        let mut lo = 0;
        let mut hi = cumfreqs.len();

        while lo < hi {
            let mid = (lo + hi) / 2;
            if value < cumfreqs[mid] {
                hi = mid;
            } else {
                lo = mid + 1;
            }
        }

        // Clamp to valid symbol range
        lo.min(cumfreqs.len() - 1)
    }

    /// Renormalization loop for decoder.
    fn renormalize(&mut self) {
        loop {
            if (self.low ^ self.high) & HALF_RANGE == 0 {
                // E1/E2: MSBs match - shift
                self.low = (self.low << 1) & STATE_MASK;
                self.high = ((self.high << 1) & STATE_MASK) | 1;
                self.code = ((self.code << 1) & STATE_MASK) | self.read_bit() as u64;
            } else if (self.low & !self.high & QUARTER_RANGE) != 0 {
                // E3: Underflow - matches CUDA arith_cuda.cu
                self.low = ((self.low << 1) ^ HALF_RANGE) & STATE_MASK;
                self.high = (((self.high ^ HALF_RANGE) << 1) | HALF_RANGE | 1) & STATE_MASK;
                // CUDA: code = (code & HALF_RANGE) | ((code << 1) & (STATE_MASK >> 1)) | bit
                // This preserves the MSB, shifts rest left, and adds new bit
                let half_bit = self.code & HALF_RANGE;
                let shifted = (self.code << 1) & (STATE_MASK >> 1);
                self.code = half_bit | shifted | self.read_bit() as u64;
            } else {
                break;
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_uniform_cumfreqs(n: usize) -> Vec<u32> {
        // Uniform distribution: each symbol has equal frequency
        let freq_each = 1000u32;
        (1..=n).map(|i| i as u32 * freq_each).collect()
    }

    #[test]
    fn test_encode_decode_single() {
        let cumfreqs = make_uniform_cumfreqs(4);

        let mut encoder = ArithEncoder::new();
        encoder.encode_symbol(&cumfreqs, 2);
        encoder.finish();

        let encoded = encoder.get_output().to_vec();
        let mut decoder = ArithDecoder::new(encoded);
        let decoded = decoder.decode_symbol(&cumfreqs);

        assert_eq!(decoded, 2);
    }

    #[test]
    fn test_encode_decode_sequence() {
        let cumfreqs = make_uniform_cumfreqs(8);
        let symbols = vec![0, 3, 7, 2, 5, 1];

        let mut encoder = ArithEncoder::new();
        for &sym in &symbols {
            encoder.encode_symbol(&cumfreqs, sym);
        }
        encoder.finish();

        let encoded = encoder.get_output().to_vec();
        let mut decoder = ArithDecoder::new(encoded);

        for &expected in &symbols {
            let decoded = decoder.decode_symbol(&cumfreqs);
            assert_eq!(decoded, expected);
        }
    }

    #[test]
    fn test_encode_decode_longer_sequence() {
        // Test with a longer sequence to verify stability
        let cumfreqs = make_uniform_cumfreqs(8);
        let symbols: Vec<usize> = (0..16).map(|i| i % 8).collect();

        let mut encoder = ArithEncoder::new();
        for &sym in &symbols {
            encoder.encode_symbol(&cumfreqs, sym);
        }
        encoder.finish();

        let encoded = encoder.get_output().to_vec();
        let mut decoder = ArithDecoder::new(encoded);

        for &expected in &symbols {
            let decoded = decoder.decode_symbol(&cumfreqs);
            assert_eq!(decoded, expected);
        }
    }

    #[test]
    fn test_reset() {
        let cumfreqs = make_uniform_cumfreqs(4);

        let mut encoder = ArithEncoder::new();
        encoder.encode_symbol(&cumfreqs, 1);
        encoder.finish();

        encoder.reset();
        encoder.encode_symbol(&cumfreqs, 2);
        encoder.finish();

        let encoded = encoder.get_output().to_vec();
        let mut decoder = ArithDecoder::new(encoded);
        assert_eq!(decoder.decode_symbol(&cumfreqs), 2);
    }
}
